coffeescript-loader/loader.js

```diff
@@ -25,41 +34,50 @@ export function resolve(specifier, context, defaultResolve) {
+export async function load(url, context, defaultLoad) {
   // Now that we patched resolve to let CoffeeScript URLs through, we need to
   // tell Node.js what format such URLs should be interpreted as. Because
   // CoffeeScript transpiles into JavaScript, it should be one of the two
   // JavaScript formats: 'commonjs' or 'module'.
   if (extensionsRegex.test(url)) {
-    // CoffeeScript files can be either CommonJS or ES modules, so we want any
-    // CoffeeScript file to be treated by Node.js the same as a .js file would
-    // be at the same location (based on the "type" field in the nearest
-    // parent package.json file). To determine how Node.js would interpret an
-    // arbitrary .js file, append .js to our CoffeeScript URL and see what
-    // format Node.js returns for such a URL. If 'commonjs' is returned, a
-    // handler will need to be registered with require.extensions to process
-    // that file via the CommonJS loader.
-    const { format } = defaultGetFormat(`${url}.js`);
-    return {format};
-  }
+    // CoffeeScript files can be either CommonJS or ES modules, but since load
+    // handles both format and source based on the same url, it cannot be used
+    // for non-js files. Instead, a quick search up the filesystem for a
+    // package.json with a `type` field settles the format issue.
+    const format = await getPackageType(dirname(fileURLToPath(url)));
 
-  // Let Node.js handle all other URLs.
-  return defaultGetFormat(url, context, defaultGetFormat);
-}
+    // source is ignored (never checked) for cjs, so safe to omit
+    if (format === 'commonjs') return { format };
 
-export function transformSource(source, context, defaultTransformSource) {
-  const { url, format } = context;
+    const { source: rawSource } = await defaultLoad(url, { format });
+    // This hook converts CoffeeScript source code into JavaScript source code
+    // for all imported CoffeeScript files.
+    const transformedSource = CoffeeScript.compile(''+rawSource, {
+      bare: true,
+      filename: url,
+    });
 
-  // This hook converts CoffeeScript source code into JavaScript source code
-  // for all imported CoffeeScript files.
-  if (extensionsRegex.test(url)) {
     return {
-      source: CoffeeScript.compile(source, {bare: true, filename: url})
+      format,
+      source: transformedSource,
     };
   }
 
-  // Let Node.js handle all other sources.
-  return defaultTransformSource(source, context, defaultTransformSource);
+  // Let Node.js handle all other URLs.
+  return defaultLoad(url, context, defaultLoad);
+}
+
+function getPackageType(dir) {
+  const packagePath = resolvePath(dir, 'package.json');
+
+  return readFile(packagePath, { encoding: 'utf8' })
+    .then((filestring) => JSON.parse(filestring))
+    .then(({ type }) => type)
+    .catch((err) => {
+      if (err?.code !== 'ENOENT') console.error(err);
+
+      return dir.length > 1 && getPackageType(resolvePath(dir, '..'))
+    });
```

Comment:
Why not just call `readFile` in a `try`/`catch` block, and save the filesystem call? We should also write this in `async`/`await` style. Something like (untested):

```suggestion
function getPackageType(url) {
  const containingFolder = dirname(fileURLToPath(url));
  try {
    const packageMetadata = await readFile(join(containingFolder, 'package.json'), 'utf-8');
  } catch {
    if (containingFolder.length > 1) {
      return getPackageType(join(containingFolder, '..', 'package.json'));
    } else {
      // We’ve reached the root of the file system and didn’t find a package.json
      return 'commonjs';
    }
  }
  return JSON.parse(packageMetadata).type || 'commonjs';
```

Reply:
Hmm? How does a `try`/`catch` avoid an FS call? If you're thinking there should be a check on `containingFolder.length`, it already does that:

```js
// .catch
return dir.length > 1 && getPackageType(resolvePath(dir, '..'))
```

If `dir.length > 1` fails, it short-circuits and settles with boolean `false` (instead of recursing again, effectively avoiding the predestined failing fs call I think you're concerned about). I did that intentionally so it would fail a subsequent typecheck (as well as a truthy/falsy check), causing Node to alert the user that something has gone wrong (whereas returning a valid value could effectively silence a problem, where the user is expecting their value to be found). I would ideally settle with `undefined` (since literally no value was defined), but that's more verbose to implement, so I opted for succinct-and-close-enough.

Try / catch is ugly and awful (and forces more nesting). I would only await when I'm fairly certain whatever I'm awaiting will not throw (and something above catches it), or i would attach a `.catch` to the awaited call. I'm not so keen to provide an example I myself would reject in a PR. I also find returning the promise chain to be much cleaner—it clearly says "the final result of this will be the output" (whereas try/catch can have difficult-to-notice early returns, and deeply nested early returns are especially difficult to notice for a poorly sighted person).

If you feel realllly strongly about the try/catch 😬

Reply:
A few things: the filesystem call I was thinking of was `resolvePath`; does that not hit the filesystem? I’m not that familiar with `path.resolve`, so I’d believe you if you told me it didn’t. I’m pretty sure that `path.join` just does string concatenation without any file I/O, which is why I was preferring it.

As for `try`/`catch`, I don’t feel strongly about that; though I’m not sure why `.catch` is significantly better? I’m just thinking we should follow the style of the docs, and I think the newest examples in the docs try to generally use `await` as much as possible instead of the older Promise chain or callback styles. If there’s a way to use `await` without `try`/`catch`, sure, do that; otherwise, why would you prefer a promise chain over `await` with `try`/`catch`? At the very least I think you should collapse the two `.then` calls, since the `JSON.parse` isn’t returning a promise: `.then(fileContents => JSON.parse(fileContents).type || 'commonjs')`.

You also need to account for the case where you find a `package.json` but it doesn’t have a `type` field, which is treated as if it had `"type": "commonjs"`.

Reply:
As I understood it, the only difference between `join` and `resolve` is that `resolve` will prepend the normalised path with the cwd when the result is otherwise not an absolute path (whereas `join` would just return the unprepended path segment), and I think the prepending behaviour is desirable here juuust in case (if `path.join` resulted in a non-absolute path, I think the correct version of that would be the cwd-prepended value that `path.resolve` does).

I'm not sure if `process.cwd()` is actually an fs call (I would guess Node would cache this and update its cache on change).

Aside from that, I just checked both the [win32](https://github.com/nodejs/node/blob/master/lib/path.js#L158) and [posix implementations](https://github.com/nodejs/node/blob/master/lib/path.js#L1085), and `path.resolve` looks to be merely string manipulation 🙂

> At the very least I think you should collapse the two `.then` calls, since the `JSON.parse` isn’t returning a promise

Ahh yes. I was thinking of fetch's `rsp.json()`, which does return a promise.

> You also need to account for the case where you find a `package.json` but it doesn’t have a `type` field, which is treated as if it had `"type": "commonjs"`

Ahh, yes. Thanks!